# Basic Stock Price Returns and Variance Analysis 

Completed for my "Analyzing Financial Data using Python" Codecademy course.

Robert Hall

12/25/2024

Disclaimer: Stock prices, returns, and resulting financial statistics do not reflect real stock prices, returns or resulting financial statistics for either Amazon or Ebay. Figures below were provided by Codecademy in the project template. Any similarity to real data -- past, present or future -- is entirely coincidental. 

In [12]:
from utils import *
from math import log, sqrt

In [13]:
amazon_prices = [1699.8, 1777.44, 2012.71, 2003.0, 1598.01, 1690.17, 1501.97, 1718.73, 1639.83, 1780.75, 1926.52, 1775.07, 1893.63]
ebay_prices = [35.98, 33.2, 34.35, 32.77, 28.81, 29.62, 27.86, 33.39, 37.01, 37.0, 38.6, 35.93, 39.5]

### 1. Calculate Rate(s) of Return

In [14]:
# function to calculate percentage returns given a list of monthly stock prices
def get_returns(prices):
  '''
  prices:   a python list of sequential monthly stock prices
  returns:  a list of corresponding sequential percentage monthly returns 
  '''
  returns = []
  for p in range(len(prices)-1):
    returns.append(calculate_log_return(prices[p], prices[p+1]))
  return returns

In [15]:
# returns for Amazon and eBay stock prices
amazon_returns = get_returns(amazon_prices)
print(amazon_returns)

ebay_returns = get_returns(ebay_prices)
print(ebay_returns)

[0.044663529768886545, 0.12430794584153733, -0.004836016009131401, -0.22588695153690044, 0.056070010445170376, -0.11805153581831997, 0.13480806622777397, -0.046993068074800755, 0.082442045949722, 0.07868064267475429, -0.0818754077815861, 0.06465576316168306]
[-0.080413352599944, 0.034052142745915476, -0.04708855595763511, -0.1287909136142863, 0.02772725974321574, -0.061257840487993175, 0.18106448560390354, 0.10293169244250136, -0.00027023375384007574, 0.042334363826560736, -0.07167967534535787, 0.09472807078164892]


In [25]:
import pandas as pd
months = [m for m in range(1, 13)]
df_amazon = pd.DataFrame({'Month': months,
                          'Returns': [display_as_percentage(r) for r in amazon_returns]}).set_index('Month')
df_ebay = pd.DataFrame({'Month': months,
                        'Returns': [display_as_percentage(r) for r in ebay_returns]}).set_index('Month')

#### Amazon Returns: 

In [26]:
df_amazon

Returns
Month        
1        4.5%
2       12.4%
3       -0.5%
4      -22.6%
5        5.6%
6      -11.8%
7       13.5%
8       -4.7%
9        8.2%
10       7.9%
11      -8.2%
12       6.5%

#### Ebay Returns:

In [27]:
df_ebay

Returns
Month        
1       -8.0%
2        3.4%
3       -4.7%
4      -12.9%
5        2.8%
6       -6.1%
7       18.1%
8       10.3%
9       -0.0%
10       4.2%
11      -7.2%
12       9.5%

In [28]:
amazon_annual_return = display_as_percentage(sum(amazon_returns))
ebay_annual_return = display_as_percentage(sum(ebay_returns))
print(f"Annual Amazon Return:   {amazon_annual_return}")
print(f"Annual Ebay Return:     {ebay_annual_return}")

Annual Amazon Return:   10.8%
Annual Ebay Return:     9.3%


### 2. Evaluate Variance and Standard Devation of Amazon, Ebay Returns

In [30]:
amazon_var = calculate_variance(amazon_returns)
ebay_var = calculate_variance(ebay_returns)

print(f"Variance in Amazon Returns:     {display_as_percentage(amazon_var)}")
print(f"Variance in Ebay Returns:       {display_as_percentage(ebay_var)}")

Variance in Amazon Returns:     1.1%
Variance in Ebay Returns:       0.7%


In [31]:
amazon_stdev = calculate_stddev(amazon_returns)
ebay_stdev = calculate_stddev(ebay_returns)

print(f"Std. Deviation of Amazon Returns:     {display_as_percentage(amazon_stdev)}")
print(f"Std. Deviation of Ebay Returns:       {display_as_percentage(ebay_stdev)}")

Std. Deviation of Amazon Returns:     10.4%
Std. Deviation of Ebay Returns:       8.6%


### 3. Evaluate Correlation Between Amazon and Ebay Rates

In [37]:
r = round(calculate_correlation(amazon_returns, ebay_returns), 2)
r2 = round(r ** 2, 2)
print(f"r coefficient:      {r}")
print(f"r^2 coefficient:    {r2}")

r coefficient:      0.68
r^2 coefficient:    0.46


### Summary and Key Takeaways:

* Amazon's stock price shows an annual return of 10.8%, with Ebay showing a comparably impressive 9.3%
* There exists higher variability in Amazon's monthly returns compared to Ebay's
* There exists a moderately strong relationship between Amazon's returns and Ebay's returns
* Approximately 46% of the variability in Ebay's returns can be explained by Amazon's returns